### PR TITLE
Phase2 muon validation

### PR DESCRIPTION
**Pull request by calabria**
This PR implements the muon validation/DQM for the Phase2 upgrade full reco. It is a temporary solution until all the detectors are ready making possible to run the full validation/DQM sequences.
@kpedro88 @jshlee

### CODE REVIEW

**Comment by makortel** [comment]
(This is essentially the same comment I made in the RECO meeting last Wednesday to @kpedro88's talk)

I would suggest to either (without any powers so feel free to ignore)
* use standard DQM+validation sequences and use eras to customize them for phase2
  * this approach was/is used for phase1
* add phase2-specific DQM+validation sequences and modify DQM, validation, and relval matrices accordingly

instead of ending up with several `fooOnly` sequences that are mixed together in relval matrix. That is of course unless "muon-only" DQM+validation makes otherwise sense. But even then I'd say it could be more useful/general to combine the phase2 validation sequences into a single sequence within validation configuration rather than combine them in relval matrix.

(for "trackingOnly" we have a specific use case, and the sequences were just "exploited" for phase2 as a simple starting point).

File: SimTracker/TrackAssociatorProducers/python/quickTrackAssociatorByHits_cfi.py [hunk 1]
```diff
@@ -20,3 +20,6 @@
     quickTrackAssociatorByHits.associateStrip = False
     quickTrackAssociatorByHits.associatePixel = False
     quickTrackAssociatorByHits.useClusterTPAssociation = False
+
+eras.phase2_muon.toModify( quickTrackAssociatorByHits, pixelSimLinkSrc = cms.InputTag("simSiPixelDigis","Pixel") )
+eras.phase2_muon.toModify( quickTrackAssociatorByHits, stripSimLinkSrc = cms.InputTag("simSiPixelDigis","Tracker") )
```
Comment (makortel) [inline SimTracker/TrackAssociatorProducers/python/quickTrackAssociatorByHits_cfi.py]: I believe these should be `phase2_tracker` as in `MuonAssociatorByHits_cfi.py` and `muonAssociatorByHitsNoSimHitsHelper_cfi.py` above.

(Are you running a configuration without `tpClusterProducer`? That is the only way I can see these would have any effect on top of what is done in tracker customize functions https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/SLHCUpgradeSimulations/Configuration/python/phase2TkTilted.py#L336. But yes, given this fallback mechanism https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/SimTracker/TrackAssociatorProducers/plugins/QuickTrackAssociatorByHitsProducer.cc#L166 it is good to keep these parameters in `quickTrackAssociatorByHits` and `tpClusterProducer` in synch. Maybe next time I get hit by the fallback I just remove it.)

(personally I would use `pixelSimLinkSrc = "simSiPixelDigis:Pixel"`, i.e. avoid explicit type for an existing parameter, to benefit from clone/toModify's typo checker)

Comment (calabria) [inline SimTracker/TrackAssociatorProducers/python/quickTrackAssociatorByHits_cfi.py]: Yes, you are right it is not needed, but in the beginning for some reason it was complaining and I added those lines also in quickTrackAssociatorByHits_cfi.py. I reverted the file to its original version. or do you want to keep it?

**Comment by calabria** [comment]
@makortel I understand your point, I can move to the one of the proposed solutions (one validation sequence for the phase2 upgrade or the use of eras to customize the normal sequences). Let's hear also from @kpedro88

File: SimTracker/TrackAssociatorProducers/python/quickTrackAssociatorByHits_cfi.py (hunk 1 above)
Comment (makortel) [inline SimTracker/TrackAssociatorProducers/python/quickTrackAssociatorByHits_cfi.py]: I'm fine with either way. Probably at some point the `quickTrackAssociatorByHits` parameters need to be changed though (IIRC the main user of the parameters is FastSim).

**Comment by kpedro88** [comment]
@makortel - at the moment, the HCAL validation code does not work with new geometries. Rather than frequently changing the Era customizations in numerous files to remove/modify various processes that are still a work in progress for Phase2, I would prefer just to include those sequences that are actually ready. Phase2 is not as far along as Phase1 yet.

**Comment by makortel** [comment]
@kpedro88 I admit I didn't get which of 3 options (this PR + the 2 in my comment) was your preference, so let me elaborate a bit my view in general (to give more food for thought if nothing else).

It may be a matter of taste, but with the way of this PR, in addition to the "numerous files" (`globalValidation_cff.py` and `postValidation_cff.py` in this case) one had to edit validation matrix (`autoValidation.py`) and relval matrix (`relval_steps.py`).

I mean, in any case one has to define the sequences somehow somewhere. For the sequences alone, I don't see much added complexity in using eras for customizing the standard sequences over defining new sequences. But with new sequences one has to add stuff to DQM and validation matrix files and relval matrix (i.e. more work), and if the sequences are combined in relval matrix, the `cmsDriver` line becomes longer. And if the era is already given to `cmsDriver`, why the fact that workflow is phase2 be repeated in the validation sequence specification? (already keeping era, GT, and geometry arguments in sync is a bit too much for my taste, but ok, `runTheMatrix` takes care of it) Why should validation sequence be treated differently from e.g. reconstruction sequence?

**Comment by kpedro88** [comment]
The development of Phase2 is proceeding asynchronously, which has been the plan since the beginning of the year. There are inevitable fluctuations concerning which versions of which subdetectors are in which workflows. To handle this, we have been changing the geometries, Era compositions (i.e. which sub-eras are included), and workflow definitions as necessary.

Your suggestion would require synchronized edits of the validation sequences for *all* subdetectors, either disabling or customizing them to work in the current Phase2 setup. Frankly, I don't have the time, knowledge, or interest to do this. (The process of creating a full reco workflow was bad enough, and in my limited experience with the validation code, it can be much more inflexible.) The method used in this PR lets each subdetectors' experts commission and enable their own validation sequence for Phase2, distributing the work. Any specific customizations needed for Phase2 should still be made with Eras (as they are here), but actually enabling the process (once it works) is better done in the workflow configuration. It requires more book-keeping, but that's the price we pay to allow asynchronous development.

**Comment by makortel** [comment]
@kpedro88 Fair enough. The current approach requires synchronized edits in `relval_steps.py` though, and I don't see how one could fully avoid conflicts. (and I still think that one could achieve the same level of development parallelism with standard sequences, but agree it would take some time to set things up nicely)

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
@cesare here is the error:
```
----- Begin Fatal Exception 18-Jul-2016 22:07:53 CEST-----------------------
An exception of category 'Error: uninitialized ProxyBase used' occurred while
   [0] Processing run: 1 lumi: 1 event: 2
   [1] Running path 'dqmoffline_1_step'
   [2] Calling event method for module MuonEnergyDepositAnalyzer/'muonEnergyDepositAnalyzer'
----- End Fatal Exception -------------------------------------------------
```
Interestingly it doesn't appear until event 2.

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by calabria** [comment]
In pre8 I did not have any problem, but in the latest IB the validation is giving that error... Since I don't know if the plots produced by that analyzer are meaningful or not with our implementation I prefered to remove the module for the moment.

**Comment by cvuosalo** [comment]
+1

For #15215 

Temporary implementation of muon DQM & validation for Phase 2 full reco.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_1_X_2016-07-18-2300 show no significant differences.

**Comment by dmitrijus** [comment]
+1

**Comment by kpedro88** [comment]
@civanch, @fabozzi, @srimanob, @hengne please take a look

**Comment by civanch** [comment]
+1

**Comment by fabozzi** [comment]
+1

**Comment by davidlange6** [comment]
@dmitrijus - what are the dqm group plans for avoiding the sort of constructs introduced by this PR for phase2? stuff like @baseValidation+@trackingOnlyValidation+@muonOnlyValidation,DQM:@trackingOnlyDQM+@muon

should not be exposed to runTheMatrix but rather handled internally to the dqm/validation offs

**Comment by davidlange6** [comment]
+1
to summarize the plan from the ORP - the dqm group will move this complexity into the appropriate cff with the next opportunity.

**Pull request merged**